Floating point accuracy control.

zahiraam opened the pull request:
This patch implements the accuracy controls for floating-point math functions in DPC++. Using the -ffp-accuracy command line option, the user can request an accuracy level for all math functions or for specific ones. Calls to fpbuiltin intrinsics llvm.fpbuilin.* are then generated.

Syntax: 

Linux:   -ffp-accuracy=[default|value][:funclist]
Windows: /Qfp-accuracy:[default|value][:funclist]

funclist is an optional comma separated list of math library functions.

-ffp-accuracy=[default|value]
   default: Use the implementation defined accuracy for all math library functions.
            This is equivalent to not using this option.         
   value:   Use the defined standard accuracy for what each accuracy value 
            means for all math library functions.

-ffp-accuracy=[default|value][:funclist]

   default: Use the implementation defined accuracy for the math library functions in funclist.
            This is equivalent to not using this option.
   value:   Use the defined standard accuracy for what each accuracy value 
            means for the math library functions in funclist.

   value is one of the following values denoting the library function accuracy.

high	This is equivalent to max-error = 1.0.
medium	This is equivalent to max-error = 4.
low	This is equivalent to accuracy-bits = 11 for single-precision functions. 
accuracy-bits = 26 for double-precision functions.
     sycl         Determined by the OpenCL specification for math function accuracy: https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_C.html#relative-error-as-ulps
     cuda         Determined by standard https://docs.nvidia.com/cuda/cuda-c-programming-guide/#mathematical-functions-appendix

andykaylor commented on clang/include/clang/Driver/Options.td:
```diff
@@ -1661,6 +1661,11 @@ def ffp_exception_behavior_EQ : Joined<["-"], "ffp-exception-behavior=">, Group<
+def ffp_accuracy_EQ : Joined<["-"], "ffp-accuracy=">, Group<f_Group>, Flags<[CC1Option]>,
+  HelpText<"Specifies the required accuracy for floating-point operations and library calls.">,
+  Values<"default,high,medium,low">, NormalizedValuesScope<"LangOptions">,
+  NormalizedValues<["FPA_Default", "FPA_High", "FPA_Medium", "FPA_Low"]>,
```
Why are sycl and cuda not here?

andykaylor commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -472,14 +475,63 @@ static Value *EmitISOVolatileStore(CodeGenFunction &CGF, const CallExpr *E) {
+    // on the function (IntrinsicID) and the command line accuracy (FPAccuracyIntrinsicID)
+    // FPAccuracyVal = TLI.getFPAccuracy(F->getName(), FPAccuracyIntrinsicID)
+    // auto *AccuracyMDS = MDString::get(CGF.Builder.getContext(), FPAccuracyVal);
+    return CGF.Builder.CreateCall(F, {Src0, AccuracyMD});
```
I know the mock-up I shared used a metadata argument with a string like you have hear, but I ended up switching that to represent the accuracy as a numeric value using a callsite attribute. See PR #8134.

zahiraam replied:
Yes, I knew that was temporary. I was shooting for this IR:

  call float @llvm.experimental.fpaccuracy.cos.f32(float %0, "float ulp-value")

but we need to have a way to calculate the ulp error. From you message in Team and comments below, it looks like the FE needs to compute this ulp error without going through the TargetLibraryInfoImp? So, it would be a new function depending on the library function name and the fp-accuracy option given in the command line (or FPA_default if no option is used)?

Do you mean fpbuiltin-max-error? This attribute still needs the ulp error value. 

So, this is the IR you are expecting?
  call float @llvm.experimental.fpaccuracy.cos.f32(float %0) #10
  attributes #0 = { "fpbuiltin-max-error"="2.5" }

If yes, the FE needs to generate a new attribute to mark the function and calculate the ulp error value (the way I describe above).

andykaylor replied:
I think the mapping still needs to come from the backend because we'll want other languages (like Fortran) to be able to use the same functionality. I'll think about this and try to come up with a proposal. If you want to put a placeholder function in the front end for now, that's fine.

andykaylor commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -472,14 +475,63 @@ static Value *EmitISOVolatileStore(CodeGenFunction &CGF, const CallExpr *E) {
+  if (FPAccuracyIntrinsicID != Intrinsic::not_intrinsic) {
+    LangOptions::FPAccuracyKind FPAccuracy = CGF.getLangOpts().getFPAccuracy();
+    // Enter this part of the condition only when TLI.isFPACCuracyAvailable is
+    // true, implying FPAccuary != LangOptions::FPA_Default.
```
I don't think this assumption is correct. It's possible that we'll have a back-end that supports multiple accuracy implementations but we want to use the default accuracy anyway. In fact, that's probably going to be the most common case.

andykaylor commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -472,14 +475,63 @@ static Value *EmitISOVolatileStore(CodeGenFunction &CGF, const CallExpr *E) {
-                                unsigned ConstrainedIntrinsicID) {
+                                unsigned ConstrainedIntrinsicID,
+    unsigned FPAccuracyIntrinsicID = Intrinsic::not_intrinsic) {
+  llvm::TargetLibraryInfoImpl TLII(CGF.getTarget().getTriple());
```
I don't think this will work. We'll need to populate the alternate math library info in the createTLII() function in BackendUtil.cpp. What you'll get this way will be a different TLII instance that doesn't have the alternate math library info populated.

zahiraam replied:
mh!  I don't see how FE can use TLII after the BackendUtil.cpp has populated with alternate functions? We need another mechanism.

andykaylor commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -2189,6 +2241,8 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
   // TODO: Extend this handling to all builtin calls that we can constant-fold.
   Expr::EvalResult Result;
+
+  llvm::TargetLibraryInfoImpl TLII(getTarget().getTriple());
```
As mentioned above, this won't work.

mdtoguchi commented on clang/include/clang/Basic/DiagnosticDriverKinds.td:
```diff
@@ -61,6 +61,10 @@ def err_drv_no_cuda_libdevice : Error<
+def warn_function_fp_accuray_already_set : Warning <"FP accuracy value of '%0' has already "
+  "been assigned to function '%1'">;
+def warn_all_fp_accuray_already_set : Warning <"FP accuracy value of '%0' has already "
+  "been assigned to all functions in the program">;
```
Should there be a way to disable these diagnostics?

mdtoguchi replied:
It is an informative diagnostic, but typically warnings have a way to be disabled.

zahiraam replied:
I can create the warning in a new group:
def FPAccuracyWrongValue  : DiagGroup<"wrong-value-of-fp-accuracy">;
and then use the -Wno-wrong-value-of-fp-accuracy on the command line to disable the warning. Is that what you mean?

mdtoguchi replied:
Yup, that's what I mean.  Maybe:  "fp-accuracy-value" would be sufficient here for the name.

mdtoguchi commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -5981,6 +5981,27 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
+      FpAccuracyAttr += " ";
+    FpAccuracyAttr += optStr.str();
+  };
+  for (const Arg *A : Args) {
```
Can you use `Args.getAllArgValues(options::OPT_ffp_accuracy_EQ)` here instead?

zahiraam commented:
@mdtoguchi Thanks for taking a look at it.

mdtoguchi commented on clang/include/clang/Basic/DiagnosticDriverKinds.td:
```diff
@@ -61,6 +61,10 @@ def err_drv_no_cuda_libdevice : Error<
   "via '--cuda-path', or pass '-nocudalib' to build without linking with "
   "libdevice">;
 
+def warn_function_fp_accuray_already_set : Warning <"FP accuracy value of '%0' has already "
```
```suggestion
def warn_function_fp_accuracy_already_set : Warning <"FP accuracy value of '%0' has already "
```

mdtoguchi commented on clang/include/clang/Basic/DiagnosticDriverKinds.td:
```diff
@@ -61,6 +61,10 @@ def err_drv_no_cuda_libdevice : Error<
 
+def warn_function_fp_accuray_already_set : Warning <"FP accuracy value of '%0' has already "
+  "been assigned to function '%1'">;
+def warn_all_fp_accuray_already_set : Warning <"FP accuracy value of '%0' has already "
```
```suggestion
def warn_all_fp_accuracy_already_set : Warning <"FP accuracy value of '%0' has already "
```

andykaylor commented on clang/test/CodeGen/fp-accuracy.c:
```diff
@@ -0,0 +1,46 @@
+  // CHECK-FUNC: call float @llvm.fpbuiltin.cos.f32(float {{.*}})   [[ATTR3:#[0-9]+]]
+  // CHECK-FUNC: call double @llvm.fpbuiltin.sin.f64(double {{.*}}) [[ATTR3:#[0-9]+]]
+  // CHECK-FUNC: call double @llvm.fpbuiltin.tan.f64(double {{.*}}) [[ATTR4:#[0-9]+]]
+  // CHECK-FUNC: call double @llvm.fpbuiltin.cos.f64(double {{.*}})
```
Why is the fpbuiltin intrinsic generated in this case? If it is generated, the test should do something to verify that no attribute is attached.

andykaylor commented on clang/test/CodeGen/fp-accuracy.c:
```diff
@@ -0,0 +1,46 @@
+  // CHECK: call double @llvm.fpbuiltin.sin.f64(double {{.*}})      [[ATTR3:#[0-9]+]]
+  // CHECK: call double @llvm.fpbuiltin.tan.f64(double {{.*}})      [[ATTR3:#[0-9]+]]
+  // CHECK: call double @llvm.fpbuiltin.cos.f64(double {{.*}})      [[ATTR3:#[0-9]+]]
+
```
Are you missing a check for the sincos call in the normal case?

andykaylor commented on clang/test/CodeGen/fp-accuracy.c:
```diff
@@ -0,0 +1,46 @@
+  c = sin(a);
+  d = tan(c);
+  e = cos(d);
+#ifndef SPIR
```
Why is sincos excluded for SPIR?

andykaylor commented on llvm/include/llvm/IR/FPAccuracy.def:
```diff
@@ -0,0 +1,66 @@
+//
+// FP_ACCURACY(<IID>, <SYCL_FLOAT>, <SYCL_DOUBLE>, <CUDA_FLOAT>, <CUDA_DOUBLE>)
+//
+FP_ACCURACY(fpbuiltin_fadd,   "0.0f",  "0.0f",  "0.0f",  "0.0f")
```
I wasn't expecting the 'f' on the end of the string, but if APFloat parses that correctly it's probably better.

andykaylor commented on clang/include/clang/Driver/Options.td:
```diff
@@ -1700,6 +1700,13 @@ def ffp_exception_behavior_EQ : Joined<["-"], "ffp-exception-behavior=">, Group<
+  Values<"default,high,medium,low,sycl,cuda">, NormalizedValuesScope<"LangOptions">,
+  NormalizedValues<["FPA_Default", "FPA_High", "FPA_Medium", "FPA_Low", "FPA_Sycl", "FPA_Cuda"]>,
+  MarshallingInfoEnum<LangOpts<"FPAccuracy">, "FPA_Default">;
+def ffp_accuracy_attr_EQ : Joined<["-"], "ffp-accuracy-attr=">, Group<f_Group>, Flags<[CC1Option]>;
```
Having attr in the cc1 option name is exposing an implementation detail. I think something like ffp-builtin-accuracy would be better.

andykaylor commented on clang/include/clang/Driver/Options.td:
```diff
@@ -6946,6 +6953,11 @@ class CLRemainingArgsJoined<string name> : Option<["/", "-"], name,
+def _SLASH_Qfp_accuracy_EQ : CLJoined<"Qfp-accuracy=">,
+  Alias<ffp_accuracy_EQ>;
+def _SLASH_Qfp_accuracy_COL : CLJoined<"Qfp-accuracy:">,
+  Alias<ffp_accuracy_EQ>,HelpText<"Defines the accuracy for math library "
```
For the Linux option the help string is "Specifies the required accuracy for floating-point operations and library calls". That's better and should be used here.

andykaylor commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -489,13 +490,33 @@ static Value *EmitISOVolatileStore(CodeGenFunction &CGF, const CallExpr *E) {
                                 const CallExpr *E, unsigned IntrinsicID,
                                 unsigned ConstrainedIntrinsicID) {
   llvm::Value *Src0 = CGF.EmitScalarExpr(E->getArg(0));
+  if (CGF.CGM.getCodeGenOpts().FPAccuracy) {    
```
Can you do something here to see if there is an accuracy requirement for this function? It's better to not use the fpbuiltin intrinsic if it isn't needed?

zahiraam replied:
Can you please explain why? I will add a comment.

andykaylor commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -2483,7 +2509,7 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
       return RValue::get(emitUnaryMaybeConstrainedFPBuiltin(*this, E,
                                    Intrinsic::sin,
-                                   Intrinsic::experimental_constrained_sin));
+                                   Intrinsic::fpbuiltin_sin));
```
Currently this patch is only handling sin, cos, tan, and sincos, right? We'll need to add support for the other functions in the fpbuiltin table at some point.

zahiraam replied:
Yes. Will do.

andykaylor commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -1836,6 +1837,45 @@ static bool HasStrictReturn(const CodeGenModule &Module, QualType RetTy,
+    llvm::StringSet<> FuncOwnAttrs;
+    StringRef FPAccuracyVal = llvm::fp::getAccuracyForFPBuiltin(
+        ID, FuncType, convertFPAccuracy(M.second));
+    FuncAttrs.addAttribute("fpbuiltin-max-error=", FPAccuracyVal);
```
You should check here for getAccuracyForFPBuiltin returning an empty StringRef. Also on line 1873 below.

zahiraam replied:
Okay. Using the wrong value of fp-accuracy is diagnosed early.  I can't find a case where we can arrive here with a null string.

andykaylor commented on clang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -3565,6 +3583,65 @@ void CompilerInvocation::GenerateLangArgs(const LangOptions &Opts,
+          Diags.Report(diag::err_drv_unsupported_option_argument)
+              << "ffp-accuracy" << FPAccuracy;
+        std::pair<LangOptions::FPAccuracyMapTy::iterator, bool> Result =
+            Opts.FPAccuracyMap.insert({"fp-accuracy", FPAccuracy.str()});
```
I don't understand what FPAccuracyMap is mapping. What is the difference between this and FPAccuracyFuncMap?

zahiraam replied:
FPAccuracyMap creates a map between fp-accuracy and the value of accuracy given of the command line.
For example: clang -c -ffp-accuracy=high t1.c
will create FPAccuracyMap with:
[fp-accuracy -> high]

FPAccuracyFuncMap will create a map between the functions on the command line and an FPAccuracyMap.
For example: clang -c -ffp-accuracy=high:[f1,f2] -ffp-accuracy=low:[f3] t1.c
will create a FPAccuracyFunMap with:
[f1 -> [fp-accuracy ->high]]
[f2 -> [fp-accuracy ->high]]
[f3 -> [fp-accuracy ->low]]

andykaylor replied:
So is there ever more than one entry in FPAccuracyMap?

andykaylor commented on clang/test/Driver/fp-accuracy.c:
```diff
@@ -0,0 +1,20 @@
+// RUN: %clang -### -target x86_64 -ffp-accuracy=low:sin,cos -c %s 2>&1 \
+// RUN:   | FileCheck --check-prefix=CHECK-FUNC-1 %s
+
+// RUN: %clang -### -target x86_64 -ffp-accuracy=low:sin,cos -ffp-accuracy=high:tan -c %s 2>&1 \
```
I should also be able to combine a general accuracy requirement with a function-specific accuracy requirement. For instance "-ffp-accuracy=high -ffp-accuracy=low:sin"

zahiraam commented on clang/test/CodeGen/fp-accuracy.c:
```diff
@@ -0,0 +1,46 @@
+// RUN: | FileCheck %s
+
+// RUN: %clang_cc1 -triple x86_64-unknown-unknown \
+// RUN: "-ffp-accuracy-attr=high:[sin,cosf] low:[tan] medium:[sincos]" \
```
You mean the spaces inside the string argument? Yes, that's fine.

andykaylor commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -506,18 +504,34 @@ static CallInst *CreateBuiltinCallWithAttr(CodeGenFunction &CGF,
-    Function *Func =
-        CGF.CGM.getIntrinsic(ConstrainedIntrinsicID, Src0->getType());
-    return CreateBuiltinCallWithAttr(CGF, Func, {Src0},
+  if (CGF.CGM.getCodeGenOpts().FPAccuracy) {
```
We need some way to handle the case where fp-accuracy is combined with a mode that requires constrained intrinsics. We should probably reject any such combinations in the driver, but it can also happen with pragmas, so we'll need to check it in the pragma handling too.

Eventually I'd like to have the fpbuiltin intrinsics take rounding mode and exception behavior attributes, but that is beyond the current scope.

zahiraam replied:
> We need some way to handle the case where fp-accuracy is combined with a mode that requires constrained intrinsics. We should probably reject any such combinations in the driver, but it can also happen with pragmas, so we'll need to check it in the pragma handling too.
> 
Can you give an example please?
> Eventually I'd like to have the fpbuiltin intrinsics take rounding mode and exception behavior attributes, but that is beyond the current scope.
That's fine. We could do that. These attributes can be added in to the FPAccuracyMap, in which case it will have more than one entry.

andykaylor commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -506,18 +504,34 @@ static CallInst *CreateBuiltinCallWithAttr(CodeGenFunction &CGF,
+    StringRef Name = CGF.CGM.getContext().BuiltinInfo.getName(BuiltinID);
+    Function *Func;
+    // Use fpbuiltin intrinsic only when needed.
+    bool IsBuiltin = false;
```
The use of "builtin" here could be confusing, since this function is being called in the context of handling clang builtins. I was going to suggest that using "fpbuiltin" would be better, but it's only marginally better since a clang builtin for a math function is a kind of floating-point builtin. I can't think of a better way to make the distinction concisely though. Maybe HasAccuracyRequirement?

andykaylor commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -1857,20 +1857,20 @@ void CodeGenModule::getDefaultFunctionFPAccuracyAttributes(
     StringRef FPAccuracyVal = llvm::fp::getAccuracyForFPBuiltin(
         ID, FuncType, convertFPAccuracy(M.second));
-    FuncAttrs.addAttribute("fpbuiltin-max-error=", FPAccuracyVal);
+    if (!FPAccuracyVal.empty())
```
Should we issue a diagnostic in this case (and the same check below)? An assertion might be better because I think this will only happen if the front end thinks accuracy should be available for this function but it isn't. I'm not sure what the best behavior is here, but it feels like a problem in our error handling somewhere.

zahiraam replied:
May be adding an assertion is the safest. Adding it.

andykaylor commented on clang/test/CodeGen/fp-accuracy.c:
```diff
@@ -1,46 +1,67 @@
+  // CHECK-FUNC-1: call double @llvm.fpbuiltin.sin.f64(double {{.*}}) [[ATTR_HIGH]]
+  // CHECK-FUNC-1: call double @llvm.fpbuiltin.tan.f64(double {{.*}}) [[ATTR_LOW:#[0-9]+]]
+  // CHECK-FUNC-1: call double @llvm.cos.f64(double {{.*}})
+  // CHECK-FUNC-1: call double @sincos(double {{.*}})                 [[ATTR_MEDIUM:#[0-9]+]]
```
I hadn't noticed this before. This needs to be a call to @llvm.fpbuiltin.sincos.f64(). I guess that's not happening because sincos isn't a clang builtin so it doesn't get the same handling as sin, cos, and tan. We need to make this replacement somewhere.

zahiraam replied:
The intrinsic is defined here: https://github.com/intel/llvm/blob/sycl/llvm/include/llvm/IR/Intrinsics.td#L891
but it requires 2 arguments. And as far as I know it takes only one?

andykaylor replied:
No, it's a weird function. My current definition is wrong though. It actually takes three arguments -- the input value, an output pointer for sin, and an output pointer for cos.

https://man7.org/linux/man-pages/man3/sincos.3.html

andykaylor reviewed (commented):
I've updated PR #9167 to align with your changes here.

zahiraam commented:
FPAccuracyMap has one entry, but has the possibility to have more if we add other attributes.

andykaylor commented:
> FPAccuracyMap has one entry, but has the possibility to have more if we add other attributes.

If we added other attributes, the variable would need to be renamed. I think it's better not to use a map until it is needed.

andykaylor commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -1861,6 +1861,8 @@ void CodeGenModule::getDefaultFunctionFPAccuracyAttributes(
         ID, FuncType, convertFPAccuracy(M.second));
     if (!FPAccuracyVal.empty())
       FuncAttrs.addAttribute("fpbuiltin-max-error=", FPAccuracyVal);
+    else
```
I don't like the "else assert" here. I'm not sure it will even compile correctly in non-assert builds.
```suggestion
   assert(!FPAccuracyVal.empty() && "A valid accuracy value is expected");
   FuncAttrs.addAttribute("fpbuiltin-max-error", FPAccuracyVal);
```

zahiraam commented:
> > FPAccuracyMap has one entry, but has the possibility to have more if we add other attributes.
> 
> If we added other attributes, the variable would need to be renamed. I think it's better not to use a map until it is needed.

Okay. Working on this now. Fixed the sincos issue. Thanks.

zahiraam commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -489,18 +490,56 @@ static Value *EmitISOVolatileStore(CodeGenFunction &CGF, const CallExpr *E) {
+    bool HasAccuracyRequirement = false;
+    if (!CGF.getLangOpts().FPAccuracyVal.empty())
+      HasAccuracyRequirement = true;
+    for (auto F : CGF.getLangOpts().FPAccuracyFuncMap) {
```
In LangOptions.h I have added 
std::string FPAccuracyVal;
and
FPAccuracyFuncMapTy FPAccuracyFuncMap;
First use:
clang -ffp-accuracy=Val
That means all math functions in the code will have an accuracy level of value "Val". Only FPAccuracyVal will be set. FPAccuracyFuncMap will not be filled.

Second use:
clang -ffp-accuracy=Val:[f1,f2] 
That means the functions f1 and f2 will have an accuracy level of Val.  In this case only  FPAccuracyFuncMap will be set where f1 -> Val and f2->Val.

Third use:
clang -ffp-accuracy=Val1 -ffp-accuracy=Val2:[f1,f2]
That means all functions will have an accuracy level of Val1, and f1 and f2 will have an accuracy level of Val2. In this case a warning is used and Val1 will be used as accuracy level.  In this case both FPAccuracyVal and FPAccuracyFuncMap will be set.

In the loop we are going through the map looking for the function we are processing and if it has an accuracy level mapped to it, then we can mark it as replaceable by an fpbultin intrinsic.

premanandrao replied:
But you are not using the 'F' in the loop.  Wouldn't you be able to find Name.str() in the Map without the for loop?

zahiraam replied:
Yes!

zahiraam commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -1836,6 +1837,42 @@ static bool HasStrictReturn(const CodeGenModule &Module, QualType RetTy,
+    assert(!FPAccuracyVal.empty() && "A valid accuracy value is expected");
+    FuncAttrs.addAttribute("fpbuiltin-max-error=", FPAccuracyVal);
+  }
+  if (!getLangOpts().FPAccuracyFuncMap.empty()) {
```
Fixed it.

zahiraam commented:
@intel/dpcpp-cfe-reviewers, @intel/dpcpp-clang-driver-reviewers can you please review? Thanks.

elizabethandrews commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -488,18 +489,72 @@ static Value *EmitISOVolatileStore(CodeGenFunction &CGF, const CallExpr *E) {
+        Diags.Report(E->getBeginLoc(),
+                     diag::err_incompatible_fp_accuracy_options);
+      } else {
+        unsigned BuiltinID = CGF.getCurrentBuiltinID();
```
I feel like this else has some unnecessary temporary variables used just once.  We can just call the function directly instead of using a variable here.

elizabethandrews commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -488,18 +489,72 @@ static Value *EmitISOVolatileStore(CodeGenFunction &CGF, const CallExpr *E) {
+        StringRef Name = CGF.CGM.getContext().BuiltinInfo.getName(BuiltinID);
+        Function *Func;
+        // Use fpbuiltin intrinsic only when needed.
+        bool HasAccuracyRequirement = hasAccuracyRequirement(CGF, Name);
```
Same here

elizabethandrews commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -488,18 +489,72 @@ static Value *EmitISOVolatileStore(CodeGenFunction &CGF, const CallExpr *E) {
+        Function *Func;
+        // Use fpbuiltin intrinsic only when needed.
+        bool HasAccuracyRequirement = hasAccuracyRequirement(CGF, Name);
+        Func = getIntrinsic(CGF, Src0, FPAccuracyIntrinsicID, IntrinsicID,
```
Function *Func =

elizabethandrews commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -508,12 +563,23 @@ static Value *emitUnaryMaybeConstrainedFPBuiltin(CodeGenFunction &CGF,
   llvm::Value *Src1 = CGF.EmitScalarExpr(E->getArg(1));
-
+  if (CGF.CGM.getCodeGenOpts().FPAccuracy) {
+    unsigned BuiltinID = CGF.getCurrentBuiltinID();
```
Same comments as above

elizabethandrews commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -21823,6 +21891,138 @@ RValue CodeGenFunction::EmitIntelFPGAMemBuiltin(const CallExpr *E) {
+    // If the function has a clang builtin but doesn't have an
+    // fpbuiltin, it will be generated with no 'fpbuiltin-max-error'
+    // attribute.
+    return nullptr;
```
Is this supposed to correspond to the default case in switch above?

zahiraam replied:
It could also be the case where the function doesn't have a clang builtin but is marked with a NoBuiltInAttr.

zahiraam replied:
Yes. The function that catches this control low is tanf.

elizabethandrews commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -1836,6 +1837,50 @@ static bool HasStrictReturn(const CodeGenModule &Module, QualType RetTy,
+      FuncAttrs.addAttribute("fpbuiltin-max-error=", FPAccuracyVal);
+    }
+  }
+  if (!getLangOpts().FPAccuracyVal.empty()) {
```
Should this be an else if?

zahiraam replied:
No. We need to check both. The first if is to treat the setting of the fp-accuracy with no specific functions, the second if is to treat the values of accuracy for specific functions. And in the command line we can have both: -ffp-accuracy=va1l -ffp-accuracy=val2:[f1,f2].

elizabethandrews replied:
I'm a little confused. IIUC, if `FPAccuracyFuncMap` is not empty in first `if`, you are adding to `FuncAttrs`. In this `if ` however, you are checking that the size of  `FuncAttrs` is 0 before updating it, which means you won't add the attribute again right? Am I misunderstanding something?

elizabethandrews reviewed (commented):
@zahiraam can this be implemented in clang community instead?

hchilama commented on clang/include/clang/Basic/DiagnosticDriverKinds.td:
```diff
@@ -140,6 +146,9 @@ def err_drv_incompatible_unwindlib : Error<
   "--rtlib=libgcc requires --unwindlib=libgcc">;
 def err_drv_incompatible_options : Error<
   "The combination of '%0' and '%1' is incompatible">;
+def err_drv_incompatible_fp_accuracy_options : Error<
```
Another thought,  How about warn( instead of error,) and ignore (-fmath-errno) to enable (fp-accuracy)?
Or, alternatively, ignore (fp-accuracy) to enable math-errno?

zahiraam replied:
There was a big debate about generating an error or a warning. After some discussion the error solution was adopted with that explicit error message.

hchilama replied:
ok

hchilama commented on clang/test/Driver/fp-accuracy.c:
```diff
@@ -0,0 +1,56 @@
+// RUN: -ffp-accuracy=goo:[tan] %s 2>&1  \
+// RUN: | FileCheck %s --check-prefixes=ERR-1
+
+// RUN: not %clang -Xclang -verify -fno-math-errno -ffp-accuracy=high=[sin] %s 2>& 1 \
```
is it "=" or ":" after "high"?

zahiraam commented:
> @zahiraam can this be implemented in clang community instead?
Andy proposed it to community, but without a library implementation to go with the implementation, there were more questions than support. It will eventually be accepted there, but we need a complete implementation, including at least one open-source library with multiple accuracy levels before it can be proposed again to community.

elizabethandrews commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -21823,6 +21887,136 @@ RValue CodeGenFunction::EmitIntelFPGAMemBuiltin(const CallExpr *E) {
+  llvm::CallInst *CI = nullptr;
+  StringRef Name;
+  if (CurrentBuiltinID == 0) {
+    // Even if the current function doesn't have a clang builtin, create
```
What happens if we have `NoBuiltinAttr`

elizabethandrews commented:
> > @zahiraam can this be implemented in clang community instead?

> >> Andy proposed it to community, but without a library implementation to go with the implementation, there were more questions than support. It will eventually be accepted there, but we need a complete implementation, including at least one open-source library with multiple accuracy levels before it can be proposed again to community.

Ok. I am not familiar with floating point arithmetic, so I am not sure if I am a good reviewer for functionality. I am hoping @premanandrao continues reviewing this. Since we expect this to be eventually up-streamed in community clang, can @erichkeane / @AaronBallman also take a look?

AaronBallman commented on clang/include/clang/Basic/CodeGenOptions.def:
```diff
@@ -374,6 +374,8 @@ CODEGENOPT(VirtualFunctionElimination, 1, 0) ///< Whether to apply the dead
                                              /// virtual function elimination
                                              /// optimization.
 
+CODEGENOPT(FPAccuracy, 1, 0)
```
Needs comments explaining what the option does.

AaronBallman commented on clang/include/clang/Basic/DiagnosticDriverKinds.td:
```diff
@@ -63,6 +63,12 @@ def err_drv_no_cuda_libdevice : Error<
   "via '--cuda-path', or pass '-nocudalib' to build without linking with "
   "libdevice">;
 
+def warn_function_fp_accuracy_already_set : Warning <"FP accuracy value of '%0' has already "
+  "been assigned to function '%1'">,
+  InGroup<FPAccuracyAlreadySet>;
+def warn_all_fp_accuracy_already_set : Warning <"FP accuracy value of '%0' has already "
+  "been assigned to all functions in the program">,
+  InGroup<FPAccuracyAlreadySet>;
```
FP is fine in comments or discussion, but let's spell it out in user-facing contexts like diagnostics and documentation.

Also, these can be combined with a `%select`.

zahiraam replied:
The diag at line 69 is actually a left over from previous patch. Removed it.

AaronBallman commented on clang/include/clang/Basic/DiagnosticDriverKinds.td:
```diff
@@ -140,6 +146,9 @@ def err_drv_incompatible_unwindlib : Error<
 def err_drv_incompatible_options : Error<
   "The combination of '%0' and '%1' is incompatible">;
+def err_drv_incompatible_fp_accuracy_options : Error<
+  "fp-accuracy requirements cannot be guaranteed when math-errno is enabled. "
+  "Use -fno-math-errno to enable fp-accuracy control.">;
```
This doesn't match our usual diagnostic requirements. 1) reword fp-accuracy as above, 2) math-errno should be `'-fmath-errno'` 3) change to: `... is enabled; use '-fno-math-errno' to ... ` and drop the trailing full stop.

AaronBallman commented on clang/include/clang/Basic/DiagnosticFrontendKinds.td:
```diff
@@ -60,6 +60,9 @@ def err_incompatible_fp_eval_method_options : Error<
     "%select{'fapprox-func'|'mreassociate'|'freciprocal'}0">;
 
+def err_incompatible_fp_accuracy_options : Error<
+    "option 'ffp-accuracy' cannot be used with option 'fmath-errno'">;
```
Don't we already have a more general "option X cannot be used with option Y" diagnostic instead of hard-coding the strings?

AaronBallman replied:
There's `err_drv_incompatible_options` in the driver if you want to move it to DiagnosticCommonKinds.td and use it, but I don't insist (I thought we had a general one already in use, but we don't -- that one is only used once and it's got issues with diagnostic quality that I'm going to fix upstream).

AaronBallman commented on clang/include/clang/Basic/LangOptions.h:
```diff
@@ -22,6 +22,7 @@
 #include "clang/Basic/TargetCXXABI.h"
 #include "clang/Basic/Visibility.h"
 #include "llvm/ADT/FloatingPointMode.h"
+#include "llvm/ADT/MapVector.h"
```
I don't see a map-vector being used?

AaronBallman commented on clang/include/clang/Basic/LangOptions.h:
```diff
@@ -509,6 +519,10 @@ class LangOptions : public LangOptionsBase {
   std::string OptRecordFile;
 
+  std::string FPAccuracyVal;
+  using FPAccuracyFuncMapTy = std::map<std::string, std::string>;
```
StringMap?

zahiraam replied:
hm! that doesn't seem to be working. Generating build errors. Doesn't it take StringRef and I am using std::string?

AaronBallman replied:
It does, but I'm not certain what about this requires the allocations used by std::string, either.

zahiraam replied:
Left it as is for now. Will find out more about it.

AaronBallman commented on clang/include/clang/Driver/Options.td:
```diff
@@ -6978,6 +6985,12 @@ class CLRemainingArgsJoined<string name> : Option<["/", "-"], name,
 // (We don't put any of these in cl_compile_Group as the options they alias are
 // already in the right group.)
 
+def _SLASH_Qfp_accuracy_EQ : CLJoined<"Qfp-accuracy=">,
```
Intel-specific markings since this won't be upstreamed?

AaronBallman commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -488,18 +489,70 @@ static Value *EmitISOVolatileStore(CodeGenFunction &CGF, const CallExpr *E) {
+                                           unsigned ID) {
+  llvm::CallInst *CI = CGF.Builder.CreateCall(FPBuiltinF, Args);
+  llvm::AttributeList AttrList;
+  if (Name == "sincos")
```
Why is this special?

AaronBallman commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -488,18 +489,70 @@ static Value *EmitISOVolatileStore(CodeGenFunction &CGF, const CallExpr *E) {
+  return CI;
+}
+
+Function *getIntrinsic(CodeGenFunction &CGF, llvm::Value *Src0,
```
`static`?

AaronBallman commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -488,18 +489,70 @@ static Value *EmitISOVolatileStore(CodeGenFunction &CGF, const CallExpr *E) {
+             : CGF.CGM.getIntrinsic(IntrinsicID, Src0->getType());
+}
+
+bool hasAccuracyRequirement(CodeGenFunction &CGF, StringRef Name) {
```
`static`?

AaronBallman commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -1836,6 +1837,50 @@ static bool HasStrictReturn(const CodeGenModule &Module, QualType RetTy,
          Module.getLangOpts().Sanitize.has(SanitizerKind::Return);
 }
 
+llvm::fp::FPAccuracy convertFPAccuracy(StringRef FPAccuracy) {
```
`static`?

AaronBallman replied:
Also, why do we need to do this dance using strings instead of `FPAccuracyKind` enumeration values?

zahiraam replied:
This is specific to function. So we need to find if the accuracy for each function is needed. Unless I am not understanding your comment?
What would we do for a run command like this one:
 "-ffp-builtin-accuracy=high:[acosf,cos,pow] low:[tan] medium:[sincos,log10]" 
if we used the FPAccuracyKind.

zahiraam replied:
Removed the comment :) Responded to the wrong comment.

zahiraam replied:
Not sure what you mean here? The input the function is a string, and we are returning an FPAccuracyKind.

AaronBallman commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -5557,8 +5602,15 @@ RValue CodeGenFunction::EmitCall(const CGFunctionInfo &CallInfo,
+  llvm::CallBase *CI = nullptr;
   if (!InvokeDest) {
+    if (CGM.getCodeGenOpts().FPAccuracy) {
+      const FunctionDecl *FD = dyn_cast_or_null<FunctionDecl>(TargetDecl);
```
```suggestion
      const auto *FD = dyn_cast_or_null<FunctionDecl>(TargetDecl);
```

AaronBallman reviewed (commented):
Mostly high-level comments at this point, still thinking about it.

zahiraam commented:
https://github.com/intel/llvm/pull/8280#discussion_r1198358341

zahiraam commented:
> > > @zahiraam can this be implemented in clang community instead?
> 
> > > > Andy proposed it to community, but without a library implementation to go with the implementation, there were more questions than support. It will eventually be accepted there, but we need a complete implementation, including at least one open-source library with multiple accuracy levels before it can be proposed again to community.
> 
> Ok. I am not familiar with floating point arithmetic, so I am not sure if I am a good reviewer for functionality. I am hoping @premanandrao continues reviewing this. Since we expect this to be eventually up-streamed in community clang, can @erichkeane / @AaronBallman also take a look?

Not anytime soon though!

AaronBallman commented on clang/include/clang/Basic/DiagnosticCommonKinds.td:
```diff
@@ -299,6 +299,9 @@ def warn_stack_clash_protection_inline_asm : Warning<
   InGroup<DiagGroup<"slh-asm-goto">>;
+
+def err_drv_incompatible_options : Error<
+  "The combination of '%0' and '%1' is incompatible">;
```
```suggestion
  "the combination of '%0' and '%1' is incompatible">;
```

AaronBallman commented on clang/include/clang/Basic/DiagnosticFrontendKinds.td:
```diff
@@ -60,6 +60,9 @@ def err_incompatible_fp_eval_method_options : Error<
     "%select{'fapprox-func'|'mreassociate'|'freciprocal'}0">;
 
+def err_incompatible_fp_accuracy_options : Error<
+    "option 'ffp-accuracy' cannot be used with option 'fmath-errno'">;
```
```suggestion
    "option '-ffp-accuracy' cannot be used with option '-fmath-errno'">;
```
Why this and not reuse `err_drv_incompatible_options`?

AaronBallman commented on clang/include/clang/Basic/DiagnosticGroups.td:
```diff
@@ -1445,3 +1445,6 @@ def ReadOnlyPlacementChecks : DiagGroup<"read-only-types">;
 def UnsafeBufferUsage : DiagGroup<"unsafe-buffer-usage">;
+
+// Warnings for fp accuracy already set.
+def FPAccuracyAlreadySet : DiagGroup<"fp-accuracy-value">;
```
This can be removed and you can use an ad hoc group in the definition of `warn_function_fp_accuracy_already_set`.

zahiraam replied:
What do you mean but an "ad hoc" group?

AaronBallman replied:
Sorry, I should have spelled that out. :-D

```
def warn_foo : Warning<
  "foo is awful, how could you do such a thing?",
  InGroup<DiagGroup<"foo-is-the-worst">>;
```
this makes a one-off diagnostic group `-Wfoo-is-the-worst` and associates just this one warning with the group. It works great for times when there's only one diagnostic under the group flag.

AaronBallman commented on clang/include/clang/Basic/FPOptions.def:
```diff
@@ -26,4 +26,5 @@ OPTION(AllowReciprocal, bool, 1, NoSignedZero)
 OPTION(AllowApproxFunc, bool, 1, AllowReciprocal)
 OPTION(FPEvalMethod, LangOptions::FPEvalMethodKind, 2, AllowApproxFunc)
 OPTION(Float16ExcessPrecision, LangOptions::ExcessPrecisionKind, 2, FPEvalMethod)
+OPTION(FPAccuracy, LangOptions::FPAccuracyKind, 2, Float16ExcessPrecision)
```
This requires 3 bits to store all the values of `FPAccuracyKind` (are we missing test coverage? I would have expected that to cause a failure somewhere.)

zahiraam replied:
I changed it to be 3 bits, but not sure what testing I should be doing to catch the error? The LIT test in the Driver folder is exercising all values of FPAccuracyKind and was passing with the 2 bits width.

AaronBallman commented on clang/include/clang/Basic/LangOptions.def:
```diff
@@ -336,6 +336,7 @@ BENIGN_LANGOPT(RoundingMath, 1, false, "Do not assume default floating-point rou
 BENIGN_ENUM_LANGOPT(FPExceptionMode, FPExceptionModeKind, 2, FPE_Default, "FP Exception Behavior Mode type")
 BENIGN_ENUM_LANGOPT(FPEvalMethod, FPEvalMethodKind, 2, FEM_UnsetOnCommandLine, "FP type used for floating point arithmetic")
 ENUM_LANGOPT(Float16ExcessPrecision, ExcessPrecisionKind, 2, FPP_Standard, "Intermediate truncation behavior for floating point arithmetic")
+BENIGN_ENUM_LANGOPT(FPAccuracy, FPAccuracyKind, 2, FPA_Default, "Accuracy for floating point operations and library functions")
```
Same here as above.

AaronBallman commented on clang/include/clang/Basic/LangOptions.h:
```diff
@@ -303,6 +304,15 @@ class LangOptions : public LangOptionsBase {
+    FPA_Medium,
+    FPA_Low,
+    FPA_Sycl,
+    FPA_Cuda
```
```suggestion
    FPA_Cuda,
```
as a kindness to the next person adding to this enumeration.

AaronBallman commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -21823,6 +21890,138 @@ RValue CodeGenFunction::EmitIntelFPGAMemBuiltin(const CallExpr *E) {
+    // an NoBuiltin attribute.
+    if (!FD->hasAttr<NoBuiltinAttr>()) {
+      Name = FD->getName();
+      if (Name == "fadd")
+        FPAccuracyIntrinsicID = llvm::Intrinsic::fpbuiltin_fadd;
+      else if (Name == "fdiv")
+        FPAccuracyIntrinsicID = llvm::Intrinsic::fpbuiltin_fdiv;
+      else if (Name == "fmul")
+        FPAccuracyIntrinsicID = llvm::Intrinsic::fpbuiltin_fmul;
+      else if (Name == "fsub")
+        FPAccuracyIntrinsicID = llvm::Intrinsic::fpbuiltin_fsub;
+      else if (Name == "frem")
+        FPAccuracyIntrinsicID = llvm::Intrinsic::fpbuiltin_frem;
+      else if (Name == "sincos")
+        FPAccuracyIntrinsicID = llvm::Intrinsic::fpbuiltin_sincos;
+      else if (Name == "exp10")
+        FPAccuracyIntrinsicID = llvm::Intrinsic::fpbuiltin_exp10;
+      else if (Name == "rsqrt")
+        FPAccuracyIntrinsicID = llvm::Intrinsic::fpbuiltin_rsqrt;
+      else
+        llvm_unreachable("unexpected fpbuiltin ID");
```
You can use a `StringSwitch` here to make it more clear.

AaronBallman commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -21823,6 +21890,138 @@ RValue CodeGenFunction::EmitIntelFPGAMemBuiltin(const CallExpr *E) {
+    }
+  }
+  Func = CGM.getIntrinsic(FPAccuracyIntrinsicID, IRArgs[0]->getType());
+  if (IRArgs.size() == 1)
+    CI = CreateBuiltinCallWithAttr(*this, Name, Func, {IRArgs[0]},
+                                   FPAccuracyIntrinsicID);
+  else if (IRArgs.size() == 2)
+    CI = CreateBuiltinCallWithAttr(*this, FnPtr->getName(), Func,
+                                   {IRArgs[0], IRArgs[1]},
+                                   FPAccuracyIntrinsicID);
+  else if (IRArgs.size() == 3)
+    CI = CreateBuiltinCallWithAttr(*this, FnPtr->getName(), Func,
+                                   {IRArgs[0], IRArgs[1], IRArgs[2]},
+                                   FPAccuracyIntrinsicID);
```
You shouldn't need to do this dance -- ArrayRef has a ctor to create one from a SmallVector. And then I don't have to ask why we're using `Name` in one case and `FnPtr->getName()` for the other two. ;-)

AaronBallman commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -1836,6 +1837,50 @@ static bool HasStrictReturn(const CodeGenModule &Module, QualType RetTy,
 
+static llvm::fp::FPAccuracy convertFPAccuracy(StringRef FPAccuracy) {
+  StringRef AccuracyVal;
+  if (FPAccuracy == "high")
```
`StringSwitch` here as well (that also solves the no else after return issues).

AaronBallman commented on clang/lib/CodeGen/CodeGenFunction.h:
```diff
@@ -1597,6 +1597,28 @@ class CodeGenFunction : public CodeGenTypeCache {
+  unsigned CurrentBuiltinID = /*NotBuiltin*/ 0;
+
+public:
+  unsigned getCurrentBuiltinID() {
```
```suggestion
  unsigned getCurrentBuiltinID() const {
```

AaronBallman commented on clang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -3306,6 +3309,19 @@ void CompilerInvocation::GenerateLangArgs(const LangOptions &Opts,
+    GenerateArg(Args, OPT_ffp_builtin_accuracy_EQ, S, SA);
+  }
+
+  for (const auto &F : Opts.FPAccuracyFuncMap) {
+    SmallString<128> S;
+    S += F.second;
+    S += ':';
+    S += F.first;
+    GenerateArg(Args, OPT_ffp_builtin_accuracy_EQ, S, SA);
+  }
```
```suggestion
  for (const auto &F : Opts.FPAccuracyFuncMap)
    GenerateArg(Args, OPT_ffp_builtin_accuracy_EQ, (F.second + ":" + F.first).str(), SA);
```
Can use a Twine for this, I think.

AaronBallman commented on clang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -3306,6 +3309,19 @@ void CompilerInvocation::GenerateLangArgs(const LangOptions &Opts,
 #include "clang/Driver/Options.inc"
 #undef LANG_OPTION_WITH_MARSHALLING
 
+  if (!Opts.FPAccuracyVal.empty()) {
+    StringRef S = Opts.FPAccuracyVal;
+    GenerateArg(Args, OPT_ffp_builtin_accuracy_EQ, S, SA);
+  }
```
```suggestion
  if (!Opts.FPAccuracyVal.empty())
    GenerateArg(Args, OPT_ffp_builtin_accuracy_EQ, Opts.FPAccuracyVal, SA);
```

AaronBallman commented on clang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -3548,6 +3564,65 @@ void CompilerInvocation::GenerateLangArgs(const LangOptions &Opts,
+        Val.split(ValElement, ':');
+        // The option is of the form -ffp-accuracy=value.
+        if (ValElement.size() == 1) {
+          StringRef FPAccuracy = ValElement[0];
+          if (!(FPAccuracy.equals("default") || FPAccuracy.equals("high") ||
+                FPAccuracy.equals("low") || FPAccuracy.equals("medium") ||
+                FPAccuracy.equals("sycl") || FPAccuracy.equals("cuda")))
+            Diags.Report(diag::err_drv_unsupported_option_argument)
+                << "ffp-accuracy" << FPAccuracy;
+          Opts.FPAccuracyVal = ValElement[0].str();
```
`StringSwitch`

AaronBallman commented on clang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -3548,6 +3564,65 @@ void CompilerInvocation::GenerateLangArgs(const LangOptions &Opts,
+                    << FuncMap->second << FuncName.str();
+              }
+            } else {
+              StringRef FPAccuracy = ValElement[0];
```
`StringSwitch` as well

premanandrao commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -21823,6 +21890,119 @@ RValue CodeGenFunction::EmitIntelFPGAMemBuiltin(const CallExpr *E) {
+              .Default(-1);
+      assert(FPAccuracyIntrinsicID != -1 && "unexpected fpbuiltin ID");
+    } else {
+      return CI;
```
```suggestion
      return nullptr;
```

premanandrao commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -21823,6 +21890,119 @@ RValue CodeGenFunction::EmitIntelFPGAMemBuiltin(const CallExpr *E) {
+    }
+  }
+  Func = CGM.getIntrinsic(FPAccuracyIntrinsicID, IRArgs[0]->getType());
+  CI = CreateBuiltinCallWithAttr(*this, Name, Func, ArrayRef(IRArgs),
+                                 FPAccuracyIntrinsicID);
+  return CI;
```
```suggestion
  return CreateBuiltinCallWithAttr(*this, Name, Func, ArrayRef(IRArgs),
                                 FPAccuracyIntrinsicID);
```

And you can remove the CI definition above.

premanandrao commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -1836,6 +1837,43 @@ static bool HasStrictReturn(const CodeGenModule &Module, QualType RetTy,
+      .Case("medium", llvm::fp::FPAccuracy::Medium)
+      .Case("low", llvm::fp::FPAccuracy::Low)
+      .Case("sycl", llvm::fp::FPAccuracy::SYCL)
+      .Case("cuda", llvm::fp::FPAccuracy::CUDA);
```
What happens here if the string is none of the above?

premanandrao commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -1836,6 +1837,43 @@ static bool HasStrictReturn(const CodeGenModule &Module, QualType RetTy,
+      FuncAttrs.addAttribute("fpbuiltin-max-error=", FPAccuracyVal);
+    else
+      assert(FuncAttrs.attrs().size() == 1 &&
+             "Expected an attribute has already been mapped to the function");
```
I am not able to understand this despite multiple readings.
Can you add some comments with an example perhaps to show what each if condition does, as well as the size checks here?

zahiraam replied:
Added comment. Let me know if this is satisfactory.

premanandrao commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -5557,8 +5595,15 @@ RValue CodeGenFunction::EmitCall(const CGFunctionInfo &CallInfo,
   // Emit the actual call/invoke instruction.
-  llvm::CallBase *CI;
+  llvm::CallBase *CI = nullptr;
   if (!InvokeDest) {
```
Is this initialization needed?   All paths below assign a value to it.

premanandrao commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -2885,13 +2886,20 @@ static void RenderFloatingPointOptions(const ToolChain &TC, const Driver &D,
-      // Turning *off* -ffast-math restores the toolchain default.
-      MathErrno = TC.IsMathErrnoDefault();
+      // Turning *off* -ffast-math restores the toolchain default,
+      // unless the -fp-accuracy is used.
```
```suggestion
      // unless -fp-accuracy is used.
```

premanandrao commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -3159,8 +3167,9 @@ static void RenderFloatingPointOptions(const ToolChain &TC, const Driver &D,
       // MathErrno. However, turning *off* -ffast-math merely restores the
-      // toolchain default (which may be false).
-      MathErrno = TC.IsMathErrnoDefault();
+      // toolchain default (which may be false). Unless -fp-accuracy is used.
```
```suggestion
      // toolchain default (which may be false), unless -fp-accuracy is used.
```

premanandrao commented on clang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -2012,6 +2012,9 @@ bool CompilerInvocation::ParseCodeGenArgs(CodeGenOptions &Opts, ArgList &Args,
     }
   }
 
+  if (Arg *A = Args.getLastArg(options::OPT_ffp_builtin_accuracy_EQ))
```
```suggestion
  if (Args.getLastArg(options::OPT_ffp_builtin_accuracy_EQ))
```

premanandrao commented on clang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -3306,6 +3309,15 @@ void CompilerInvocation::GenerateLangArgs(const LangOptions &Opts,
 
+  if (!Opts.FPAccuracyVal.empty()) {
+    GenerateArg(Args, OPT_ffp_builtin_accuracy_EQ, Opts.FPAccuracyVal, SA);
+  }
```
Braces not necessary.

premanandrao commented on clang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -3306,6 +3309,15 @@ void CompilerInvocation::GenerateLangArgs(const LangOptions &Opts,
+  for (const auto &F : Opts.FPAccuracyFuncMap) {
+    GenerateArg(Args, OPT_ffp_builtin_accuracy_EQ, (F.second + ":" + F.first),
+                SA);
+  }
```
Braces not necessary.

hchilama reviewed (approved):
Driver changes LGTM

AaronBallman commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -504,18 +505,73 @@ static Value *EmitISOVolatileStore(CodeGenFunction &CGF, const CallExpr *E) {
+  llvm::AttributeList AttrList;
+  // sincos() doesn 't return a value, but it still has a type associated with
+  // it that corresponds to the operand type.
+  if (Name == "sincos")
+    CGF.CGM.getFPAccuracyFuncAttributes(Name, AttrList, ID, Args[0]->getType());
+  else
+    CGF.CGM.getFPAccuracyFuncAttributes(Name, AttrList, ID,
+                                        FPBuiltinF->getReturnType());
```
```suggestion
  CGF.CGM.getFPAccuracyFuncAttributes(Name, AttrList, ID, Name == "sincos" ? Args[0]->getType() : FPBuiltinF->getReturnType());
```
reformatted however clang-format wants, but this makes it more clear that the difference is only with the last argument to the call.

AaronBallman commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -504,18 +505,73 @@ static Value *EmitISOVolatileStore(CodeGenFunction &CGF, const CallExpr *E) {
+static bool hasAccuracyRequirement(CodeGenFunction &CGF, StringRef Name) {
+  if (!CGF.getLangOpts().FPAccuracyVal.empty())
+    return true;
+  if (!CGF.getLangOpts().FPAccuracyFuncMap.empty()) {
+    auto FuncMapIt = CGF.getLangOpts().FPAccuracyFuncMap.find(Name.str());
+    if (FuncMapIt != CGF.getLangOpts().FPAccuracyFuncMap.end())
+      return true;
+  }
+  return false;
```
```suggestion
  auto FuncMapIt = CGF.getLangOpts().FPAccuracyFuncMap.find(Name.str());
  return FuncMapIt != CGF.getLangOpts().FPAccuracyFuncMap.end();
```

AaronBallman commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -21866,6 +21933,117 @@ RValue CodeGenFunction::EmitIntelFPGAMemBuiltin(const CallExpr *E) {
+    // an NoBuiltin attribute.
+    if (!FD->hasAttr<NoBuiltinAttr>()) {
+      Name = FD->getName();
+      FPAccuracyIntrinsicID =
+          llvm::StringSwitch<unsigned>(Name)
+              .Case("fadd", llvm::Intrinsic::fpbuiltin_fadd)
+              .Case("fdiv", llvm::Intrinsic::fpbuiltin_fdiv)
+              .Case("fmul", llvm::Intrinsic::fpbuiltin_fmul)
+              .Case("fsub", llvm::Intrinsic::fpbuiltin_fsub)
+              .Case("frem", llvm::Intrinsic::fpbuiltin_frem)
+              .Case("sincos", llvm::Intrinsic::fpbuiltin_sincos)
+              .Case("exp10", llvm::Intrinsic::fpbuiltin_exp10)
+              .Case("rsqrt", llvm::Intrinsic::fpbuiltin_rsqrt)
+              .Default(-1);
+      assert(FPAccuracyIntrinsicID != -1 && "unexpected fpbuiltin ID");
```
```suggestion
      FPAccuracyIntrinsicID =
          llvm::StringSwitch<unsigned>(Name)
              .Case("fadd", llvm::Intrinsic::fpbuiltin_fadd)
              .Case("fdiv", llvm::Intrinsic::fpbuiltin_fdiv)
              .Case("fmul", llvm::Intrinsic::fpbuiltin_fmul)
              .Case("fsub", llvm::Intrinsic::fpbuiltin_fsub)
              .Case("frem", llvm::Intrinsic::fpbuiltin_frem)
              .Case("sincos", llvm::Intrinsic::fpbuiltin_sincos)
              .Case("exp10", llvm::Intrinsic::fpbuiltin_exp10)
              .Case("rsqrt", llvm::Intrinsic::fpbuiltin_rsqrt);
```
`StringSwitch` will assert if you fall off the end without a default, so this is equivalent.

AaronBallman commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -5581,6 +5620,13 @@ RValue CodeGenFunction::EmitCall(const CGFunctionInfo &CallInfo,
   llvm::CallBase *CI;
   if (!InvokeDest) {
+    if (CGM.getCodeGenOpts().FPAccuracy) {
+      const auto *FD = dyn_cast_or_null<FunctionDecl>(TargetDecl);
```
```suggestion
      const auto *FD = dyn_cast_if_present<FunctionDecl>(TargetDecl);
```
the other form is deprecated.

AaronBallman commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -6052,6 +6061,23 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
+    // attribute.
+    if (optStr.str() != "default") {
+      if (FpAccuracyAttr.empty())
+        FpAccuracyAttr = std::move(std::string("-ffp-builtin-accuracy="));
```
```suggestion
        FpAccuracyAttr = "-ffp-builtin-accuracy=";
```

AaronBallman commented on clang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -3570,6 +3580,75 @@ void CompilerInvocation::GenerateLangArgs(const LangOptions &Opts,
+                   .Case("cuda", true)
+                   .Default(false))
+            Diags.Report(diag::err_drv_unsupported_option_argument)
+                << "ffp-accuracy" << FPAccuracy;
```
```suggestion
                << "-ffp-accuracy" << FPAccuracy;
```

AaronBallman reviewed (approved):
LGTM

zahiraam commented:
@bader can this be merged in now? Thanks.

asudarsa commented:
Adding @asudarsa from @intel/dpcpp-tools-reviewers for review.

Thanks

AlexeySachkov reviewed (approved):
`IR/CMakeLists.txt` change LGTM

asudarsa commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -504,18 +505,67 @@ static Value *EmitISOVolatileStore(CodeGenFunction &CGF, const CallExpr *E) {
+  llvm::AttributeList AttrList;
+  // sincos() doesn't return a value, but it still has a type associated with
+  // it that corresponds to the operand type.
+  CGF.CGM.getFPAccuracyFuncAttributes(
```
Why do we need a AttrList here? Is there a scenario where more than one such attribute can be attached to a builtin call?

Thanks

zahiraam replied:
No, I can't think of one.

zahiraam replied:
Actually I think you are right that the AttrList can be a single attribute instead of a list. Added a TOTO comment as we discussed offline. Thanks.

asudarsa commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -504,18 +505,69 @@ static Value *EmitISOVolatileStore(CodeGenFunction &CGF, const CallExpr *E) {
+                                           ArrayRef<Value *> Args,
+                                           unsigned ID) {
+  llvm::CallInst *CI = CGF.Builder.CreateCall(FPBuiltinF, Args);
+  // TODO: Replace AttrList a single attribute. The call can only have a
```
```suggestion
  // TODO: Replace AttrList with a single attribute. The call can only have a
```

asudarsa reviewed (approved):
Minor typo. LGTM. Thanks for implementing this.

zahiraam commented:
@llvm-gatekeepers can you consider merging this please? Thanks.

dm-vodopyanov commented:
> @llvm-gatekeepers can you consider merging this please? Thanks.

@zahiraam please update the caption of PR in accordance with  https://github.com/intel/llvm/blob/sycl/sycl/doc/developer/ContributeToDPCPP.md#commit-message

dm-vodopyanov commented:
Will be merged as soon as post-commit turns green.

dm-vodopyanov commented:
Failures in post-commit:
```
Failed Tests (2):
  Clang :: CodeGen/fp-accuracy.c
  Clang :: Driver/fp-accuracy.c
```
https://github.com/intel/llvm/actions/runs/5267586087/jobs/9523035374

@zahiraam could you please address it?

aelovikov-intel commented:
> @zahiraam could you please address them?

Any updates on this? If the fix isn't ready, can we just disable the tests to fix post-commit?

zahiraam commented:
> > @zahiraam could you please address them?
> 
> Any updates on this? If the fix isn't ready, can we just disable the tests to fix post-commit?

@aelovikov-intel I was able to reproduce the issue and working on it. Yes for now, we can diable the LIT tests and I will work on it. Thanks.
